### PR TITLE
Add Performance Counter API access to kernel mode driver

### PR DESCRIPTION
**Pull request by ThadHouse**
We can use the windows perf counter API, but we don't need the manifest stored in the msquic binary. 

Still work left to do.

### CODE REVIEW

**Review by anrossi — APPROVED** [review]

File: src/bin/winkernel/driver.c [hunk 1]
```diff
@@ -29,6 +29,16 @@ MsQuicLibraryUnload(
     void
     );
 
+NTSTATUS
+MsQuicPcwStartup(
+    void
+);
```
Comment (nibanks) [inline src/bin/winkernel/driver.c:35]: nit: you have this a few places, but the `);` needs to be indented

File: src/bin/winkernel/msquicpcw.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+#pragma code_seg(push, "PAGE")
+
+EXTERN_C DECLSPEC_SELECTANY PPCW_REGISTRATION MsQuicPcwGlobal = NULL;
+
+EXTERN_C FORCEINLINE VOID
+MsQuicPcwInitRegistrationInformationGlobal(
+    __in_opt PPCW_CALLBACK Callback,
+    __in_opt PVOID CallbackContext,
+    __out PCW_REGISTRATION_INFORMATION* RegInfo
+)
```
Comment (nibanks) [inline src/bin/winkernel/msquicpcw.c:15]: nit: indent

File: src/bin/winkernel/msquicpcw.c [hunk 3]
```diff
@@ -0,0 +1,190 @@
+#include <wdm.h>
+#include <ip2string.h>
+#include "msquic.h"
+#include "quic_platform_winkernel.h"
+
+#pragma code_seg(push, "PAGE")
+
+EXTERN_C DECLSPEC_SELECTANY PPCW_REGISTRATION MsQuicPcwGlobal = NULL;
```
Comment (nibanks) [inline src/bin/winkernel/msquicpcw.c:8]: All these `EXTERN_C`s are unnecessary. Please remove.

File: src/bin/winkernel/msquicpcw.c [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+EXTERN_C FORCEINLINE VOID
```
Comment (nibanks) [inline src/bin/winkernel/msquicpcw.c:10]: Remove the `FORCEINLINE`. Replace Windows types with C types where possible (`VOID` -> `void`). Same below.

File: src/bin/winkernel/msquicpcw.c [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+MsQuicPcwInitRegistrationInformationGlobal(
```
Comment (nibanks) [inline src/bin/winkernel/msquicpcw.c:11]: Pointless to have a separate function for this. Just put this code in the caller.

File: src/bin/winkernel/msquicpcw.c [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    __in_opt PPCW_CALLBACK Callback,
```
Comment (nibanks) [inline src/bin/winkernel/msquicpcw.c:12]: Update all these SAL annotations to the new format (e.g. `_In_opt_`).

File: src/bin/winkernel/msquicpcw.c [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                &UnicodeName,
+                0,
+                Buffer);
+        break;
+    }
+
+    return Status;
+}
+
+static BOOLEAN MsQuicGlobalCountersRegistered = FALSE;
```
Comment (nibanks) [inline src/bin/winkernel/msquicpcw.c:159]: Let's put all globals at the top of the file.

File: src/bin/winkernel/msquicpcw.c [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+NTSTATUS
+MsQuicPcwStartup(
+    void
+    )
+{
+    NTSTATUS Status = STATUS_SUCCESS;
+
+    PAGED_CODE();
+
+    Status = MsQuicPcwRegisterGlobal(MsQuicPcwGlobalCallback, NULL);
```
Comment (nibanks) [inline src/bin/winkernel/msquicpcw.c:172]: Does this fail if the manifest isn't present? Let's make sure to log any functions we call that fail.

File: src/bin/winkernel/msquicpcw.c [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+    RegInfo->Version = PCW_CURRENT_VERSION;
+    RegInfo->Name = &Name;
+    RegInfo->CounterCount = RTL_NUMBER_OF(Descriptors);
+    RegInfo->Counters = (PCW_COUNTER_DESCRIPTOR*)Descriptors;
+    RegInfo->Callback = Callback;
+    RegInfo->CallbackContext = CallbackContext;
+}
+
+EXTERN_C FORCEINLINE NTSTATUS
+MsQuicPcwRegisterGlobal(
```
Comment (nibanks) [inline src/bin/winkernel/msquicpcw.c:62]: This function can also be removed and push the code to the caller.

File: src/bin/winkernel/msquicpcw.c [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+static PPCW_REGISTRATION MsQuicPcwGlobal = NULL;
+static BOOLEAN MsQuicGlobalCountersRegistered = FALSE;
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+void
+QuicLibrarySumPerfCountersExternal(
+    _Out_writes_bytes_(BufferLength) uint8_t* Buffer,
+    _In_ uint32_t BufferLength
+);
```
Comment (nibanks) [inline src/bin/winkernel/msquicpcw.c:33]: nit: indent

**Review by nibanks — APPROVED** [review]
LGTM. Minor nits.

**Pull request merged**